Source/Particles/Gather/FieldGather.H

```diff
@@ -880,6 +880,681 @@ void doGatherShapeNEsirkepovStencilImplicit (
+        sintheta_mid = 0._rt;
+    }
+    const Complex xy_mid0 = Complex{costheta_mid, sintheta_mid};
+    // Keep these double to avoid bug in single precision
+    double const x_new = (rp_new - xmin)*dxi;
+    double const x_old = (rp_old - xmin)*dxi;
+    double const x_bar = (rp_mid - xmin)*dxi;
+#elif !defined(WARPX_DIM_1D_Z)
+    // Keep these double to avoid bug in single precision
+    double const x_new = (xp_np1 - xmin)*dxi;
+    double const x_old = (xp_n - xmin)*dxi;
+    double const x_bar = (xp_nph - xmin)*dxi;
+#endif
+#if defined(WARPX_DIM_3D)
+    // Keep these double to avoid bug in single precision
+    double const y_new = (yp_np1 - ymin)*dyi;
+    double const y_old = (yp_n   - ymin)*dyi;
+    double const y_bar = (yp_nph - ymin)*dyi;
+#endif
+    // Keep these double to avoid bug in single precision
+    double const z_new = (zp_np1 - zmin)*dzi;
+    double const z_old = (zp_n   - zmin)*dzi;
+    double const z_bar = (zp_nph - zmin)*dzi;
+
+    // 1) Determine the number of segments.
+    // 2) Loop over segments and gather electric field.
+    // 3) Gather magnetic field.
+
+    // cell crossings are defined at cell edges if depos_order is odd
+    // cell crossings are defined at cell centers if depos_order is even
+
+    int num_segments = 1;
+    amrex::Real shift = 0.0_rt;
+    if ( (depos_order % 2) == 0 ) { shift = 0.5_rt; }
+
+#if defined(WARPX_DIM_3D)
+
+    // compute cell crossings in X-direction
+    const auto i_old = static_cast<int>(x_old-shift);
+    const auto i_new = static_cast<int>(x_new-shift);
+    int cell_crossings_x = std::abs(i_new-i_old);
+    num_segments += cell_crossings_x;
+
+    // compute cell crossings in Y-direction
+    const auto j_old = static_cast<int>(y_old-shift);
+    const auto j_new = static_cast<int>(y_new-shift);
+    int cell_crossings_y = std::abs(j_new-j_old);
+    num_segments += cell_crossings_y;
+
+    // compute cell crossings in Z-direction
+    const auto k_old = static_cast<int>(z_old-shift);
+    const auto k_new = static_cast<int>(z_new-shift);
+    int cell_crossings_z = std::abs(k_new-k_old);
+    num_segments += cell_crossings_z;
+
+    // need to assert that the number of cell crossings in each direction
+    // is within the range permitted by the number of guard cells
+    // e.g., if (num_segments > 5) ...
+
+    // compute total change in particle position and the initial cell
+    // locations in each direction used to find the position at cell crossings.
+    const double dxp = x_new - x_old;
+    const double dyp = y_new - y_old;
+    const double dzp = z_new - z_old;
+    const auto dirX_sign = static_cast<int>(dxp < 0. ? -1 : 1);
+    const auto dirY_sign = static_cast<int>(dyp < 0. ? -1 : 1);
+    const auto dirZ_sign = static_cast<int>(dzp < 0. ? -1 : 1);
+    double Xcell = 0., Ycell = 0., Zcell = 0.;
+    if (num_segments > 1) {
+        Xcell = i_old + shift + 0.5_rt*(1-dirX_sign);
+        Ycell = j_old + shift + 0.5_rt*(1-dirY_sign);
```

Comment:
As above: explicit case for the ints would be safer and deciding if `amre::Real/ParticleReal` or enforce `double` is the best choice.

Reply:
Got it.